marlowe/specification/ReadMe.md

```diff
@@ -0,0 +1,346 @@
+  where isMarloweAddress (Address (ScriptCredential hash)) = hash == marloweValidatorHash
+        isMarloweAddress _ = False
+```
+Hereafter we use `outValue` for the value being output to the script address and `outDatum` being the datum associated with that UTxO.
+
+
+#### *Constraint 4.* No output to script on close
+
+When a contract closes, the transaction does not output to the Marlowe script address.
+```haskell
+txOutContract transactionOutput == Close ⇒ filter (isMaroweAddress . txOutAddress) (txInfoOutputs $ scriptContextTxInfo scriptContext) ≡ []
+  where isMarloweAddress (Address (ScriptCredential hash)) = hash == marloweValidatorHash
+        isMarloweAddress _ = False
+```
+
+
+#### *Constraint 5.* Input value from script
+
+The value of the Marlowe UTxO being spent exactly matches the value recorded in the Marlowe state.
+```haskell
+inValue ≡ foldMap toValue (accounts $ marloweState marloweData)
+  where toValue ((_, Token currency name), count)) = Value.singleton currency name count
+```
+
+
+#### *Constraint 6.* Output value to script
+
+The beginning balance plus the deposits equals the ending balance plus the payments.
+```haskell
+inValue + foldMap valueOfDeposit (fmap getInputContent transactionInput) ≡ outValue + foldMap valueOfPayment (txOutPayments transactionOutput)
+  where getInputContent (Input inputContent) = inputContent
+        getInputContent (MerkleizedTxInput inputContent _) = inputContent
+        valueOfDeposit (IDeposit _ _ (Token currency name) count) = Value.singleton currency name count
+        valueOfDeposit _ = mempty
+        valueOfPayment (Payment _ (Party _) value)) = value
+        valueOfPayment _ = mempty
+```
+
+
+#### *Constraint 7.* Input state
+
+The input state of the transaction is used in the semantics computation.
+```haskell
+marloweState marloweData ≡ inState
+```
+
+
+#### *Constraint 8.* Input contract
+
+The input contract of the transaction is used in the semantics computation.
+```haskell
+marloweContract marloweData ≡ inContract
+```
+
+
+#### *Constraint 9.* Marlowe parameters
+
+The Marlowe parameters are not changed by the transaction.
+```haskell
+Just $ marloweParams marloweData ≡ fmap marloweParams . fromBuiltinData $ getDatum outDatum
+```
+
+
+#### *Constraint 10.* Output state
+
+The output state of the transaction matches the result of the semantics computation.
+```haskell
+Just $ txOutState transactionOutput ≡ fmap marloweState . fromBuiltinData $ getDatum outDatum
+```
+
+
+#### *Constraint 11.* Output contract
+
+The output contract of the transaction matches the result of the semantics computation.
+```haskell
+Just $ txOutContract transactionOutput ≡ fmap marloweContract . fromBuiltinData $ getDatum outDatum
+```
+
+
+#### *Constraint 12.* Merkleized continuations
+
+All of the merkleized contracts referenced in the input are provided to the script.
```

Comment:
provided by? how?

Reply:
fixed in 8be04323e4ec24c4e923ed214094e85c916ab8e0